flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraOutputFormatBase.java

```diff
@@ -17,130 +17,160 @@
+ * CassandraOutputFormatBase is the common abstract class for writing into Apache Cassandra using
+ * output formats.
  *
  * @param <OUT> Type of the elements to write.
  */
-public abstract class CassandraOutputFormatBase<OUT> extends RichOutputFormat<OUT> {
+public abstract class CassandraOutputFormatBase<OUT, V> extends RichOutputFormat<OUT> {
     private static final Logger LOG = LoggerFactory.getLogger(CassandraOutputFormatBase.class);
 
-    private final String insertQuery;
     private final ClusterBuilder builder;
+    private Semaphore semaphore;
+    private Duration maxConcurrentRequestsTimeout = Duration.ofMillis(Long.MAX_VALUE);
+    private int maxConcurrentRequests = Integer.MAX_VALUE;
 
     private transient Cluster cluster;
-    private transient Session session;
-    private transient PreparedStatement prepared;
-    private transient FutureCallback<ResultSet> callback;
-    private transient Throwable exception = null;
-
-    public CassandraOutputFormatBase(String insertQuery, ClusterBuilder builder) {
-        Preconditions.checkArgument(
-                !Strings.isNullOrEmpty(insertQuery), "Query cannot be null or empty");
+    protected transient Session session;
+    private transient FutureCallback<V> callback;
+    private AtomicReference<Throwable> throwable;
+
+    public CassandraOutputFormatBase(
+            ClusterBuilder builder,
+            int maxConcurrentRequests,
+            Duration maxConcurrentRequestsTimeout) {
         Preconditions.checkNotNull(builder, "Builder cannot be null");
-
-        this.insertQuery = insertQuery;
         this.builder = builder;
+        Preconditions.checkArgument(
+                maxConcurrentRequests > 0, "Max concurrent requests is expected to be positive");
+        this.maxConcurrentRequests = maxConcurrentRequests;
+        Preconditions.checkNotNull(
+                maxConcurrentRequestsTimeout, "Max concurrent requests timeout cannot be null");
+        Preconditions.checkArgument(
+                !maxConcurrentRequestsTimeout.isNegative(),
+                "Max concurrent requests timeout is expected to be positive");
+        this.maxConcurrentRequestsTimeout = maxConcurrentRequestsTimeout;
     }
 
+    /** Configure the connection to Cassandra. */
     @Override
     public void configure(Configuration parameters) {
         this.cluster = builder.getCluster();
     }
 
-    /**
-     * Opens a Session to Cassandra and initializes the prepared statement.
-     *
-     * @param taskNumber The number of the parallel instance.
-     * @throws IOException Thrown, if the output could not be opened due to an I/O problem.
-     */
+    /** Opens a Session to Cassandra . */
     @Override
     public void open(int taskNumber, int numTasks) throws IOException {
+        throwable = new AtomicReference<>();
+        this.semaphore = new Semaphore(maxConcurrentRequests);
         this.session = cluster.connect();
```

Comment:
I'm wondering if it would make sense to move the dependencies on the datastax driver into the sub-classes (columnar/pojo) at the cost of duplicating a few lines.
This would allow us to test this class in nicer fashion without mocking.

Reply:
Well, I don't like mocks either but remember the pojo and columnar (tuple and row) output formats are properly tested in `CassandraConnectorITCase`; In this ITest they actually write to a real Cassandra testContainer (that I implemented in another PR). So the interactions with the Cassandra driver are already tested there. `CassandraOutputFormatBaseTest` is just for unit tests of the flush mechanism (semaphore management, exceptions etc...) so I think a mock of the Cassandra driver which is not really needed looks fine to me. And I think that I prefer that than duplicating code.

Reply:
Alternatively we could move the session stuff into another shared sub-class.

Reply:
I agree it is a pity to use a mock only because of cassandra session/cluster and only for one test. I'll try to refactor to avoid mocking and code duplication. I'll propose you something in a isolated commit for ease of review.

Reply:
I'm thinking that if we remove everything Cassandra related from `CassandraOutputFormatBase` then this class can become the base class for all the output formats that want to leverage the flush logic. So we could promote the resulting class to `org.apache.flink.api.common.io` package as `OutputFormatBase`.  `CassandraOutputFormatBase` would extend this class and contain the cassandra deps.

I'll do the above code so that we discuss on the code itself if needed.

Reply:
done.
I introduced `OutputFormatBase` that has no Cassandra dependencies and that contains only flush and max concurrent requests features. This class is for now in cassandra module waiting for the above promotion decision. The introduction of this class is isolated in commit a3323a1b to ease a cherrypick in case we do a separate PR to address https://issues.apache.org/jira/browse/FLINK-27884